### PR TITLE
291 Allow the use of formatters without the full UI context in React / React Native

### PR DESCRIPTION
**Pull request by mimi-acn**
# what was done so far
- added dedicated formatting context provider & hook
- non-breaking migration of current UI context to use this new provider for its formatters & added deprecation note
- added basic documentation examples
- adjust react example app
- adjust any current examples of UI provider using formatters
- add more documentation on how to use formatters in React regardless of ui component library

# what still needs to be done
- indicate which formatters are problematic in React Native

# issue: 
https://github.com/bonfhir/bonfhir/issues/291

### CODE REVIEW

**Comment by opiation** [comment]
This will eventually be blocked by #293 and I will address that issue so as to allow PRs like this one to eventually get merged.

File: docs/website/packages/react/components/use-fhir-formatters-context.md
Comment (opiation) [inline docs/website/packages/react/components/use-fhir-formatters-context.md]: Nitpick: I think inlining `setLocale` reads better in docs.

e.g.:

```tsx
<button onClick={() => setLocale("en")}>EN</button>
```

It's almost imperative English, _On click, set the locale to "en"_. It also makes for shorter documentation :shrug:.

File: packages/react/src/r5/context.ts
Comment (opiation) [inline packages/react/src/r5/context.ts]: Why exactly do we remove the null check assertion that was there before?

File: packages/core/src/r5/formatters.ts
Comment (opiation) [inline packages/core/src/r5/formatters.ts]: Is it intentional that we also throw for _nullish_ values (e.g.: `""`, `false`)? If not, then we should explicitly check for `null` and `undefined`.

File: packages/react/src/r5/formatters/fhir-formatters-context.tsx
Comment (opiation) [inline packages/react/src/r5/formatters/fhir-formatters-context.tsx]: This would be a good place for a sane default

File: docs/website/packages/react/components/use-fhir-formatters-context.md
Comment (mimi-acn) [reply docs/website/packages/react/components/use-fhir-formatters-context.md]: but I like to teach proper react naming conventions 😂

File: packages/react/src/r5/context.ts
Comment (mimi-acn) [reply packages/react/src/r5/context.ts]: ah sorry, will put it back!

File: packages/core/src/r5/formatters.ts
Comment (mimi-acn) [reply packages/core/src/r5/formatters.ts]: will add explicit checks

File: packages/react/src/r5/formatters/fhir-formatters-context.tsx
Comment (mimi-acn) [reply packages/react/src/r5/formatters/fhir-formatters-context.tsx]: hmmm might only be confusing tough? Without a provider, this context is pretty much dead weight?

Comment (mimi-acn) [reply packages/react/src/r5/formatters/fhir-formatters-context.tsx]: I tried to add them, hopefully it doesn't break stuff ahahaha

File: docs/website/packages/react/components/use-fhir-formatters-context.md
Comment (opiation) [reply docs/website/packages/react/components/use-fhir-formatters-context.md]: I don't think I'd support any argument about this being a difference in naming convention apart from explicitly mentioning the _handling_ of the user action event. In addition though, I think this is mostly a difference in prose and I'm not willing to fight over style like this unless the code actually like something truly misleading. Your call.

File: packages/react/src/r4b/context.ts
Comment (opiation) [inline packages/react/src/r4b/context.ts]: The this context's initial value should change to some default UI context value that does actually conform to the `FhirUIContext` interface or its type should be widen to `FhirUIContext | undefined` with its initial value set to `undefined`.

With this change, the error a developer will get when forgetting to provide a context value above with state that some property is missing because the empty object will circumvent the null check from `useFhirUIContext`  without providing the runtime guarantee that the context value actually conforms to `FhirUIContext`.

File: packages/react/src/r4b/provider.tsx
Comment (opiation) [inline packages/react/src/r4b/provider.tsx]: Nitpick: Can we use a single noun everywhere around the formatter? I understand the utility composes a collection of type-specific formatters but the thing actually being used is a single instance of `Formatter`. I think we can also return just the formatter in that case and not an object with a `formatter` property. Unless I'm missing something, this could be:

```jsx
function Example() {
  const formatter = useFhirFormatter();

  return <>{formatter.format("HumanName", ...)}</>
}
```

File: packages/react/src/r4b/provider.tsx
Comment (opiation) [inline packages/react/src/r4b/provider.tsx:31]: Nitpick: This can use shorthand notation.

**Review by opiation — CHANGES_REQUESTED** [review]

File: packages/react/src/r4b/context.ts
Comment (mimi-acn) [reply packages/react/src/r4b/context.ts]: the problem with expecting types TContext | undefined is trailing the undefined everywhere you use the context

there is a "magic trick" I can do, so I'll submit that one, but you might not like it 😂

File: packages/react/src/r4b/provider.tsx
Comment (mimi-acn) [reply packages/react/src/r4b/provider.tsx:31]: done!

File: packages/react/src/r4b/provider.tsx
Comment (mimi-acn) [reply packages/react/src/r4b/provider.tsx]: done! had to do some juggling with names since FhirFormatter is a react component

**Review by opiation — APPROVED** [review]

**Pull request merged**
